File: quiz/quiz_engine.py
import json
import random

class QuizEngine:
    def __init__(self, questions_file="quiz/questions.json", cheeses_file="quiz/cheeses.json"):
        self.questions_file = questions_file
        self.cheeses_file = cheeses_file
        self.questions = []
        self.cheeses = []
        self.current_score = 0
        self.current_session_questions = []
        self.question_index = 0

    def load_data(self):
        """Loads questions and cheeses from JSON files."""
        try:
            with open(self.questions_file, 'r') as f:
                self.questions = json.load(f)
            with open(self.cheeses_file, 'r') as f:
                self.cheeses = json.load(f)
        except FileNotFoundError:
            print("Error: Data files not found.")
            self.questions = []
            self.cheeses = []

    def start_new_game(self, num_questions=10):
        """Resets score and picks random questions + mandatory lactose question."""
        self.current_score = 0
        self.question_index = 0
        
        # 1. Pick 10 random questions
        # Ensure we don't crash if we have fewer than num_questions
        count = min(len(self.questions), num_questions)
        self.current_session_questions = random.sample(self.questions, count)
        
        # 2. Add the mandatory Lactose Intolerant question
        lactose_question = {
            "id": 999,
            "type": "choice",
            "text": "Final Question: Are you lactose intolerant?",
            "options": [
                {"label": "Yes, sadly.", "value": "LACTOSE_YES"}, 
                {"label": "No, I am strong.", "value": 0}
            ]
        }
        self.current_session_questions.append(lactose_question)

    def get_next_question(self):
        """Returns the current question object or None if finished."""
        if self.question_index < len(self.current_session_questions):
            return self.current_session_questions[self.question_index]
        return None

    def submit_answer(self, value):
        """Adds points to score and advances index. Handles special lactose override."""
        # Check for the special lactose flag
        if value == "LACTOSE_YES":
            self.current_score = -100
        else:
            # Only add points if it's not the lactose override (which sets it directly)
            # and if the current score isn't already -100 (though this is the last question anyway)
            if isinstance(value, (int, float)):
                self.current_score += value
        
        self.question_index += 1

    def calculate_result(self):
        """Determines the cheese persona based on final score."""
        # If they are lactose intolerant, they are the Wisp (-100)
        if self.current_score == -100:
            # Find the specific Wisp entry or fallback
            for cheese in self.cheeses:
                if cheese.get("score") == -100:
                    return cheese
        
        # Normal logic: Find the closest match or range
        # The cheeses.json uses "score" as a target/min value. 
        # Let's assume we find the cheese with the highest score that is <= current_score
        # Sort cheeses by score descending to find the best match
        
        sorted_cheeses = sorted(self.cheeses, key=lambda x: x.get("score", 0), reverse=True)
        
        for cheese in sorted_cheeses:
            if self.current_score >= cheese.get("score", 0):
                return cheese
                
        # Fallback to the lowest scoring cheese if nothing matches
        return sorted_cheeses[-1] if sorted_cheeses else None

File: react/src/logic/QuizEngine.ts
import questionsData from '../../assets/questions.json';
import cheesesData from '../../assets/cheeses.json';

export interface QuestionOption {
    label: string;
    value: number | string; // number for score, string for "LACTOSE_YES" logic
}

export interface Question {
    id: number;
    type: string;
    text: string;
    options: QuestionOption[];
}

export interface Cheese {
    name: string;
    score: number;
    description: string;
    image: string;
}

export class QuizEngine {
    private questions: Question[];
    private cheeses: Cheese[];
    private currentSessionQuestions: Question[];
    private currentScore: number;
    private questionIndex: number;

    constructor(questions?: Question[], cheeses?: Cheese[]) {
        // Cast imported JSON to types (assuming JSON structure matches)
        this.questions = (questions || questionsData) as Question[];
        this.cheeses = (cheeses || cheesesData) as Cheese[];
        this.currentSessionQuestions = [];
        this.currentScore = 0;
        this.questionIndex = 0;
    }

    startNewGame(numQuestions = 10) {
        this.currentScore = 0;
        this.questionIndex = 0;

        // 1. Pick 10 random questions
        const shuffled = [...this.questions].sort(() => 0.5 - Math.random());
        this.currentSessionQuestions = shuffled.slice(0, Math.min(numQuestions, shuffled.length));

        // 2. Add Mandatory Lactose Intolerant Question
        const lactoseQuestion: Question = {
            id: 999,
            type: "choice",
            text: "Final Question: Are you lactose intolerant?",
            options: [
                { label: "Yes, sadly.", value: "LACTOSE_YES" },
                { label: "No, I am strong.", value: 0 }
            ]
        };
        this.currentSessionQuestions.push(lactoseQuestion);
    }

    getNextQuestion(): Question | null {
        if (this.questionIndex < this.currentSessionQuestions.length) {
            return this.currentSessionQuestions[this.questionIndex];
        }
        return null;
    }

    submitAnswer(value: number | string) {
        if (value === "LACTOSE_YES") {
            this.currentScore = -100;
        } else {
            if (typeof value === 'number') {
                this.currentScore += value;
            }
        }
        this.questionIndex++;
    }

    getCurrentProgress(): number {
        return this.questionIndex / this.currentSessionQuestions.length;
    }

    calculateResult(): Cheese | null {
        // 1. Handle Special Case
        if (this.currentScore === -100) {
            return this.cheeses.find(c => c.score === -100) || null;
        }

        // 2. Exact Match
        const exact = this.cheeses.find(c => c.score === this.currentScore);
        if (exact) return exact;

        // 3. Find Neighbors
        const sortedCheeses = [...this.cheeses].sort((a, b) => a.score - b.score);
        let lowerCheese: Cheese | null = null;
        let upperCheese: Cheese | null = null;

        for (const cheese of sortedCheeses) {
            if (cheese.score < this.currentScore) {
                lowerCheese = cheese;
            } else if (cheese.score > this.currentScore) {
                upperCheese = cheese;
                break;
            }
        }

        // 4. Coin Flip Logic
        if (lowerCheese && upperCheese) {
            return Math.random() < 0.5 ? lowerCheese : upperCheese;
        }

        if (lowerCheese && !upperCheese) return lowerCheese;
        if (upperCheese && !lowerCheese) return upperCheese;

        return null;
    }
}
